### PR TITLE
Add action for mikrotik routerOS

### PR DESCRIPTION
**Pull request by a16bitsysop**
Before submitting your PR, please review the following checklist:

- [ X] **CHOOSE CORRECT BRANCH**: if filing a bugfix/enhancement
      against certain release version, choose `0.9`, `0.10` or `0.11` branch,
      for dev-edition use `master` branch
- [X ] **CONSIDER adding a unit test** if your PR resolves an issue
- [ X] **LIST ISSUES** this PR resolves
- [ X] **MAKE SURE** this PR doesn't break existing tests
- [ X] **KEEP PR small** so it could be easily reviewed.
- [ X] **AVOID** making unnecessary stylistic changes in unrelated code
- [X ] **ACCOMPANY** each new `failregex` for filter `X` with sample log lines
      within `fail2ban/tests/files/logs/X` file

### CODE REVIEW

**Comment by Omniflux** [comment]
I don't think this line is correct
`actionunban = %(mtikcommand)s "/ip firewall address-list remove [find list=\"%(mtiklistname)s\" comment=%(mtikcomment)s]"`

Did you mean
`actionunban = %(mtikcommand)s "/ip firewall address-list remove [find list=\"%(mtiklistname)s\" address=<ip>]"`

For actionban, you probably want to add a timeout
`actionban = %(mtikcommand)s "/ip firewall address-list add list=\"%(mtiklistname)s\" address=<ip> comment=%(mtikcomment)s timeout=<bantime>"`

In my experience, if you have many bans, because of the setup time for the ssh connection, this causes fail2ban to take a long time when restarting or stopping because each unban requires its own connection. You probably want to add the following lines

`actionstop = %(actionflush)s`
`actionflush = %(mtikcommand)s "/ip firewall address-list remove [find list=\"%(mtiklistname)s\"]"`

**Comment by Omniflux** [comment]
I just realized you were using the comment to track the address, although it is not obvious why, so the first part of my previous comment may be incorrect.

**Comment by a16bitsysop** [comment]
Yes that is correct, it uses comment to track.  

Then you can have different jails add to the same address list if needed, but I guess doing by address would be faster with alot of bans and different address lists could be used for different jails.  Mikrotik/Router OS is supposed to be fast at finding addresses in address lists.

**Comment by sebres** [comment]
> For actionban, you probably want to add a timeout

I would not do that for several reasons, for instance to hold the time of end of ban under fail2ban control only - if, no matter why, it would be unbanned in mikrotik earlier due to internal expiration, fail2ban would still consider the ticket as banned, in this time it would ignore the next ban (as "already banned") next time a recidive intruder makes a failures again. See #2703 for more info about similar issues.

> In my experience, if you have many bans, because of the setup time for the ssh connection, this causes fail2ban to take a long time when restarting or stopping because each unban requires its own connection.

Newer version of fail2ban has a command `actionflush` to make a "bulk" unban of whole jail. This could be specified, if mikrotik has such possibility (e. g. remove all entries with comment matching `AutoF2B-<name>-*`).

**Comment by a16bitsysop** [comment]
It should be possible to do the action flush with matching comment wild card, I will edit rule and test it out

File: config/action.d/mikrotik.conf
Comment (sebres) [inline config/action.d/mikrotik.conf:51]: hmm... since `startcomment` and `mtiklistname` don't contain jail name (`<name>`), wouldn't this flush in any lists, regardless the jail? So if some jail gets stopped (restarted), it removes all tickets of other jails.

Shall not it be either:
```diff
-actionflush = %(mtikcommand)s "/ip firewall address-list remove [find list=\"%(mtiklistname)s\" comment~\"%(startcomment)s*\"]"
+actionflush = %(mtikcommand)s "/ip firewall address-list remove [find list=\"%(mtiklistname)s\" comment~\"%(startcomment)s-<name>-*\"]"
```
or the startcomment get a jail name:
```diff
-startcomment = AutoF2B
+startcomment = f2b-<name>
-mtikcomment = %(startcomment)s-<name>-<ip>
+mtikcomment = %(startcomment)s-<ip>
```

**Review by sebres — CHANGES_REQUESTED** [review]

File: config/action.d/mikrotik.conf
Comment (a16bitsysop) [reply config/action.d/mikrotik.conf:51]: Yes that looks a better way, will change to method 2

**Review by sebres — APPROVED** [review]

**Comment by vitkabele** [comment]
Is there any progress on this?

**Comment by sebres** [comment]
> Is there any progress on this?

Progress on what exactly? Can't you use the action?

I can surely merge it, just after all one has to wait for new release anyway.
As for the quality or the functioning I can't say anything since I'm not familiar with that approach, so I was hopping someone else would tell own expertise about.

**Comment by vitkabele** [comment]
Progress on merging to release. It also seems to have merge conflicts now. I'll test the action and report back whether it works for me.

**Comment by vitkabele** [comment]
It works for me, but:

1. I'd change the default parameter for `mtiklistname` to something without a space, e.g. `fail2ban`.

2. The `mtik` prefix from the parameter names could be removed (?). I am not sure what is the scope of action parameter names, however other actions seem to define parameters as if it was private scope. 

3. The `mtiklistname` should imo be just `list`.

I can't however push to the source branch, so it's up to @a16bitsysop if he wants to implement these changes.

**Comment by a16bitsysop** [comment]
sounds okay to me

File: config/action.d/mikrotik.conf
Comment (sebres) [inline config/action.d/mikrotik.conf:59]: The last adjustment made the action broken - the command parameter used in action-tags called `command`, but here it is `mtikcommand`.

File: config/action.d/mikrotik.conf
Comment (sebres) [inline config/action.d/mikrotik.conf:63]: same here and other following params (in command they are used without `mtik` prefix)

**Review by sebres — CHANGES_REQUESTED** [review]
please repair

**Comment by a16bitsysop** [comment]
missed those

**Comment by vitkabele** [comment]
Works for me with the changes 👍

**Comment by vitkabele** [comment]
@sebres Hello, it seems that the requested changes are already fixed

**Review by sebres — APPROVED** [review]

**Pull request merged**
